### PR TITLE
Record a subscription failure before advancing its read cursor

### LINKED ISSUE #1257: Subscription can drop a failed message: cursor advances before the recovery record is durable
## Summary

An asynchronous event-store subscription can **drop a failed message instead of retrying it**: it advances its read cursor before durably recording the failure to the recovery stream, so a crash (or a failure of the recovery write itself) in that window loses the message.

Found while authoring the consistency & delivery guarantees spec (#1200, sub-issue of #1199).

## Detail

In `process_batch` the cursor advance runs **before** the failure is recorded, and the durable cursor flush is not coupled to the recovery record:

```
is_successful = await self.engine.handle_message(...)   # handler ran
await self.update_read_position(position)               # cursor advances; durable every position_update_interval
...
else:  # failure
    await self._record_failed_position(position, ...)   # recovery-stream write, AFTER
```

(`src/protean/server/subscription/event_store_subscription.py` — cursor advance ~`:643`, `_record_failed_position` ~`:660-745`, durable flush at the `position_update_interval` boundary ~`:361-362`.)

## Failure windows

1. **Crash between the durable cursor advance and the recovery write.** If the failing message is the one that makes the batch counter reach `position_update_interval`, the cursor is durably persisted at that position; a crash before `_record_failed_position` writes its `Failed` record means restart resumes past the message and it is never recovered.
2. **The recovery-stream append itself fails.** `_record_failed_position` awaits `store._write(...)`; if that raises, the in-memory cursor has already advanced, and a later durable flush persists past the message.

Both produce silent loss, defeating the at-least-once/recovery contract for a *failed* message. (The batched-checkpoint at-least-once property for the happy path is unaffected.)

## Fix direction

Record the failure durably **before** advancing (and durably flushing) the cursor past a failed position, so the recovery record and the cursor cannot diverge on a crash. Add a test that injects a handler failure and a crash between handling and the recovery write, and asserts the message is retried, not skipped.

## Related
- #1088 / ADR-0025 — the `$all` no-skip fix (a different mechanism; this is the recovery-of-failed-message path).
- #1200 / #1199 — the guarantees spec that documents this caveat.

### PR DESCRIPTION
**Pull request by subhashb**
## Summary

Fixes a silent lost-message path in asynchronous event-store subscriptions.

In `EventStoreSubscription.process_batch`, when a handler failed, the read cursor was advanced (and durably checkpointed every `position_update_interval` messages) **before** the failure was recorded to the recovery stream via `_record_failed_position`. Two windows dropped the message rather than retrying it:

1. **Crash between the durable cursor advance and the recovery write.** If the failing message was the one that hit the checkpoint interval, the cursor was durably persisted past it; a crash before the `Failed` record was written meant restart resumed past the message and it was never recovered.
2. **The recovery-stream write itself fails.** `_record_failed_position` awaits `store._write`; if that raised, the cursor had already advanced, and a later durable flush persisted past the message.

Both defeated the at-least-once/recovery contract for a *failed* message. The happy-path batched-checkpoint at-least-once property was never affected.

## Fix

Reorder the per-message failure path so the durable `Failed` record is written **before** the cursor advances (and is durably flushed) past the position. Now:

- If the recovery write raises, the exception propagates **before** the cursor moves, so the message is re-read and retried on the next tick.
- If a crash lands between the record and the advance, the cursor sits at or behind the failed position with the `Failed` record already durable, so the message is re-read or recovered.

Recovery of a failed message is now crash-safe, matching the at-least-once guarantee already documented for the happy path. The change is a reorder within the loop: still exactly one `update_read_position` and at most one `_record_failed_position` per message. It is compatible with the `$all` gap-safety `tick()` two-phase watermark advance (ADR-0025): an exception in the failure path propagates and skips the watermark advance, leaving the cursor at real progress.

## Docs

- `docs/reference/guarantees.md`: the recovery caveat flips from a `!!! warning` ("can drop the failed message") to a `!!! note` ("Recovery of a failed message is crash-safe").
- The class and method docstrings now state the record-before-advance ordering.
- `changes/1257.fixed.md` changelog fragment.

## Test plan

- [x] New tests (all via a `_FaultyWrite` fault injector at a precise seam):
  - `test_recovery_write_failure_retries_message_on_next_tick` — recovery write raises → cursor unmoved, the event is re-read from the store and retried on the next `tick()`.
  - `test_failure_recorded_before_cursor_write_is_issued` — cursor write raises → the `Failed` record is already written.
  - `test_multi_message_batch_records_failure_before_durable_flush` — with a batched checkpoint (interval 2), both failures are recorded before the durable flush that raises on the later message.
  - `test_recovery_disabled_failure_advances_cursor_without_recording` — `enable_recovery=False` drops a failure by design (cursor advances, nothing recorded).
  - `test_success_records_idempotency_when_store_active` / `test_failure_does_not_record_idempotency` — the relocated `record_success` fires only on success (positive + negative), covered in core with a stub idempotency store.
- [x] Revert-check: reverting the source reorder reddens all three ordering tests (verified independently).
- [x] Core suite green (2 pre-existing flakes unrelated to this change, confirmed identical on `origin/main`); `tests/server` + `tests/subscription` + `$all` gap-safety + checkpoint-no-skip all pass.
- [x] ruff + mypy clean; 100% patch coverage.
- [x] Adversarial red-team panel (completeness, test efficacy, claim accuracy, adversarial inputs): the reorder holds on all five edge cases (stale gap watermark, in-memory/durable divergence, partial batch, duplicate side-effects, `position_update_interval > 1`); prose and test findings folded in.

Closes #1257

### CODE REVIEW

File: src/protean/server/subscription/event_store_subscription.py [hunk 1]
```diff
@@ -677,6 +678,21 @@ async def process_batch(self, messages: list[Message]) -> int:
                         stream_position=stream_position,
                     )
```
[the remark this answers is not in the document]
Comment (subhashb) [reply src/protean/server/subscription/event_store_subscription.py:679]: Good catch, fixed in 4cb945d3. Moved the `self._failed_positions[position] = {...}` update to after the durable `store._write` in `_record_failed_position`, so a failed recovery write no longer leaves a stale in-memory entry (the retry_count read stays before the write). Added an assertion to the retry test that the position is absent from `_failed_positions` after the failed write. To be precise on impact: this was never data loss (the message is still re-read and retried), but the stale entry could make the recovery pass redundantly re-deliver a message that already succeeded on a re-read, which this removes.
